### DISCUSSION TITLE
Cannot import static Image URLs

### DISCUSSION THREAD

**Discussion by vinceau** [facebook/stylex#277] [Q&A]
**The problem**
Importing static Image URLs throws `Only static values are allowed inside of stylex.create() call` error.

**How to reproduce**

Steps to reproduce:
1. Import an image, e.g.
```ts
import logo from "@/styles/images/logo.png";
```
2. Create the style with a background image.

```ts
const styles = stylex.create({
  logo: {
    backgroundImage: `url("${logo}")`,
  },
});
```
4. Use it in the component. e.g.

```ts
const Component = () => <div {...stylex.props(styles.logo)} />
```

**Expected behavior**
The image to be rendered, and the correct image URL to be inserted into the generated CSS.

**Discussion**
I understand that in the docs it says:
<img width="851" alt="Screenshot 2023-12-29 at 1 04 33 pm" src="https://github.com/facebook/stylex/assets/8384734/e3feaae0-b3a5-4e0e-914a-c59c319c0ba5">

However, the Webpack plugin should be able to resolve the path to the image (whether that's a Base64 url or otherwise), and the generated CSS should use that string. These are not run-time values and are known at compile time, so it should be possible to import this.

I understand that a workaround can be achieved by doing something like this:

```ts
const styles = stylex.create({
  logo: (logoUrl: string) => ({
    backgroundImage: `url("${logoUrl}")`,
 }),
});
```

But as I said before, these values are known at compile time and thus should be valid values.

Please advise what the recommended way of importing and setting background images in StyleX.

**Environment (include versions). Did this work in previous versions?**
StyleX version: v0.3.0
StyleX Webpack version: v0.3.0
Node version: v18.17.0
MacOS Sonoma 14.2.1

**Comment by nmn** [comment]
All styles need to be statically defined in the *same file*. So this wouldn't normally work.

However, since in this case you're importing an asset and not a file, you're relying on tooling to convert that import into a static string. If this process can be encapsulated as a Babel plugin, you can configure StyleX to resolve the import into a string *before* stylex tries to compile the file.

**Reply by nmn** [reply]
Does anyone know of a Babel plugin that can convert asset imports into URLs? I've only ever seen bundler plugins for this.

**Reply by vinceau** [reply]
So then given a React application and someone wanting to use StyleX as the CSS in JS library of choice, what is the best practice/recommendation for importing and setting background images in CSS?

Would it be to pass it in via a variable?

```tsx
import logo from "../path/to/logo.png";

const styles = stylex.create({
  logo: (logoUrl: string) => ({
    backgroundImage: `url("${logoUrl}")`,
 }),
});

const Component = () => (
  <div {...stylex.props(styles.logo(logo))} />
);
```

or would it be to just use the style prop?

```tsx
import logo from "../path/to/logo.png";

const Component = () => (
  <div {...stylex.props(styles.otherStyles)} style={{backgroundImage: `url("${logo}")`}} />
);
```

or some other suggestion you might have?

**Reply by nmn** [reply]
Of the two options, this is preferable.

```tsx
const styles = stylex.create({
  logo: (logoUrl: string) => ({
    backgroundImage: `url("${logoUrl}")`,
 }),
});
```

The ideal solution would be to use an actual URL string instead of an import or to convert the import into a URL *before* StyleX runs.

**Reply by vinceau** [reply]
How would one "convert the import into a URL before StyleX runs"?

**Reply by nmn** [reply]
This depends on the exact bundler being used. When you import assets in `webpack` and other bundlers, they are transpiled into strings. They would need to be configured to run the StyleX transform after that is done.

I'm not sure how this can be achieved without help from bundler experts.
